src/librbd/api/Mirror.cc

```diff
@@ -438,43 +442,283 @@ struct C_ImageSnapshotCreate : public Context {
 
     auto req = mirror::snapshot::CreatePrimaryRequest<I>::create(
       ictx, mirror_image.global_image_id, CEPH_NOSNAP, snap_create_flags, 0U,
-      snap_id, on_finish);
+      group_snap_id, snap_id, on_finish);
+    req->send();
+  }
+};
+
+template <typename I>
+void close_images(std::vector<I *> *image_ctxs) {
+  std::vector<C_SaferCond*> on_finishes;
+
+  for (size_t i = 0; i < image_ctxs->size(); i++) {
+    if ((*image_ctxs)[i] != nullptr) {
+      C_SaferCond* on_finish = new C_SaferCond;
+      (*image_ctxs)[i]->state->close(on_finish);
+      on_finishes.push_back(on_finish);
+    } else {
+      on_finishes.push_back(nullptr);
+    }
+  }
+
+  for (size_t i = 0; i < image_ctxs->size(); i++) {
+    if (on_finishes[i] != nullptr) {
+      on_finishes[i]->wait();
+      delete on_finishes[i];
+    }
+  }
+
+  image_ctxs->clear();
+}
+
+template <typename I>
+int open_group_images(IoCtx& group_ioctx, const std::string &group_id,
+                      std::vector<I *> *image_ctxs) {
+  CephContext *cct = (CephContext *)group_ioctx.cct();
+  std::vector<cls::rbd::GroupImageStatus> images;
+  int r = Group<I>::group_image_list_by_id(group_ioctx, group_id, &images);
+  if (r < 0) {
+    lderr(cct) << "failed group image list by id: "
+               << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  std::vector<C_SaferCond*> on_finishes;
+  for (size_t i = 0; i < images.size(); i++) {
+    auto &image = images[i];
+    librbd::IoCtx image_io_ctx;
+    r = librbd::util::create_ioctx(group_ioctx, "image", image.spec.pool_id, {},
+                                   &image_io_ctx);
+    if (r < 0) {
+      image_ctxs->push_back(nullptr);
+      on_finishes.push_back(nullptr);
+      continue;
```

Comment:
I don't think this is right -- wouldn't this just swallow the error?  Instead of opening exactly `images.size()` images, any number less than that (including 0, i.e. no images at all) can be opened but `open_group_images()` would still return success.

Reply:
Oops! Yeah, right. Hope its fine now.